### PR TITLE
 [AggressiveInstCombine] Expand memchr with small constant strings

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch converts memchr with a small constant string into a switch. It will reduce overhead of libcall and enable more folds (e.g., comparing the result with null).

References:
https://en.cppreference.com/w/c/string/byte/memchr

### CODE REVIEW

File: llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp:1127]: I would say you should either go all in on "Reading past `Str.size()` is UB" or drop this.

All in being, if `Call->getArgOperand(2) > Str.size()` we *must* find a match so the result cannot be `NULL` (drop default case in the switch). Or (IMO the better choice for now) conservatively don't do this transform if `Call->getArgOperand(2) > Str.size()`.

File: llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp
Comment (dtcxzyw) [reply llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp:1132]: Of course. But this case is rare.

Comment (goldsteinn) [reply llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp:1132]: Probabaly no strings like "aaaaasaabbbbbccccc", but there are a lot of 4/5 letter word with a dup or two

File: llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp
Comment (nikic) [inline llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp:58]: How did you chose this limit?

File: llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp
Comment (nikic) [inline llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp:1128]:
```suggestion
      // Ignore the case that n is larger than the size of string.
```

File: llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp
Comment (nikic) [inline llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp:1154]:
```suggestion
      BasicBlock::Create(Call->getContext(), "", BB->getParent(), BBNext);
```
Probably? Passing BBSuccess to it's own constructor does not make sense. Maybe this will fix the incorrect BB ordering in your tests.

Comment (nikic) [reply llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp:1154]: Also I'd give some meaningful names to the BBs, e.g. this one could be `memchr.success`.

File: llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp
Comment (dtcxzyw) [reply llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp:58]: See https://github.com/dtcxzyw/llvm-tools/blob/main/strchr.cpp

```
Len Count
3 305
4 14
5 31
6 35
7 25
8 6
9 10
10 4
11 80
12 3
13 7
14 9
15 16
16 11
17 35
18 2
19 3
20 4
22 1
23 5
24 3
25 3
29 1
33 10
37 2
38 1
41 1
45 2
49 5
53 1
54 5
56 1
63 1
64 8
65 9
66 1
78 18
86 1
91 1
```

Comment (dtcxzyw) [reply llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp:58]: After applying @goldsteinn's suggestion in https://github.com/llvm/llvm-project/pull/98501#discussion_r1675751184:
```
EffectiveLen Count
3 305
4 14
5 32
6 34
7 25
8 6
9 10
10 5
11 80
12 3
13 10
14 7
15 14
16 11
17 35
18 2
19 4
20 4
22 1
23 11
24 3
25 2
29 1
33 4
37 2
38 1
41 1
45 2
49 5
53 1
54 5
56 1
63 3
64 7
65 9
66 1
68 1
78 18
```

File: llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp:1122]: Doesn't `strchr` with a constant string + non-constant search character already get optimized to memchr:
https://github.com/llvm/llvm-project/blob/9f4a25e2a7cd176bd4f946dc651bc18c7a2e8c92/llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp#L477-L497

Can this function be made to only handle memchr?

File: llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp:1179]: Does the new code get properly optimized if the call is only used for equality?

Can you add a few tests / modify a few of your existing tests to test that case?

File: llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp:1156]: is N correct here given you can skip cases if they are a duplicate character?

File: llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp:1179]: This is also a pretty dense section of code with a lot of similarly named variables. Can you split it up a bit based on the the actual codegen order i.e

```
 PHINode *PHI = PHINode::Create(Call->getType(), 2, "", BBNext->begin());
  PHI->addIncoming(Constant::getNullValue(Call->getType()), BB);
```
and 
```
Value *FirstOccursLocation = IRB.CreateInBoundsPtrAdd(Base, IndexPHI);
PHI->addIncoming(FirstOccursLocation, BBSuccess);
```

After the loop (where they are emitted).

Likewise can you comment the different stages to help break them up a bit and decrease the code density.

File: llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp
Comment (dtcxzyw) [reply llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp:1156]: > is N correct here given you can skip cases if they are a duplicate character?

It is just a hint for memory allocation.
https://github.com/llvm/llvm-project/blob/27d961dab3caa89a78e68af094ddc8365d97dc91/llvm/include/llvm/IR/Instructions.h#L2505-L2520

File: llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp
Comment (dtcxzyw) [reply llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp:1170]: > Does the new code get properly optimized if the call is only used for equality?

No. It should be handled in InstCombine.

> Can you add a few tests / modify a few of your existing tests to test that case?

See the test `test_memchr_null`.

Comment (goldsteinn) [reply llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp:1170]: It does seem to be optimized properly by instcombine i.e: https://godbolt.org/z/o7TYnTMnT

**Comment by goldsteinn** [comment]
LGTM, wait on nikics approval or a few days to push please.

File: llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp
Comment (nikic) [inline llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp:1161]:
```suggestion
                                            BB->getParent(), BBSuccess);
```
We should insert the case blocks before the success block.

File: llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp
Comment (nikic) [inline llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp:1148]:
```suggestion
  PHINode *IndexPHI = IRB.CreatePHI(IndexTy, N, "memchr.idx");
```

File: llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp
Comment (nikic) [inline llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp:1174]: Add `PHI->takeName(Call);`

**Review by nikic — APPROVED** [review]
LGTM

**Comment by goldsteinn** [comment]
You should probably drop the `strchr` references from the commit message before pushing.

**Pull request merged**
